### ISSUE TITLE
DamageOnlyWorkflow attempts to use null item

### ISSUE THREAD

**Issue by tposney** [tposney/midi-qol#432]
In GitLab by @nickelpro on Jul 12, 2021, 21:09

Using the DamageOnlyWorkflow to append or modify damage causes a type error when the workflow tries to use `this.item` in the ROLLFINISHED state.

Simple example using an OnUse macro:

```javascript
let damageRoll = new Roll(`1d6`).roll()
new MidiQOL.DamageOnlyWorkflow(actor, token, damageRoll.total, "necrotic", args[0].hitTargets, damageRoll, {itemCardId: args[0].itemCardId})
```

Which causes the following error when run

```
workflow.js:638 Uncaught (in promise) TypeError: Cannot read property 'hasSave' of null
    at workflow.js:638
    at Map.filter (collection.mjs:66)
    at DamageOnlyWorkflow._next (workflow.js:626)
    at DamageOnlyWorkflow._next (workflow.js:1723)
    at async DamageOnlyWorkflow.next (workflow.js:334)
```

Providing the item with the `itemData` option causes other issues. It changes the damage type of the workflow and still causes a type error.

```
Uncaught (in promise) TypeError: Cannot read property 'attackBonus' of undefined
    at applyTokenDamageMany (utils.js:228)
    at Object.applyTokenDamage (utils.js:205)
    at Workflow.eval (eval at callMacro (workflow.js:866), <anonymous>:24:11)
    at Workflow.callMacro (workflow.js:869)
    at Workflow.callMacros (workflow.js:825)
    at Workflow._next (workflow.js:601)
    at Workflow.next (workflow.js:334)
    at Workflow._next (workflow.js:596)
    at Workflow.next (workflow.js:334)
    at Workflow._next (workflow.js:583)
```

**Comment by tposney** [comment]
The first issue should hav ebeen resolved in 0.8.32 - can you confirm which version you are using?

The second I will have to have a look at.

**Comment by tposney** [comment]
In GitLab by @nickelpro on Jul 12, 2021, 21:24

Ah, my bad, I'm on 0.8.31 and the auto update isn't grabbing 0.8.32 for some reason. Will close after testing with 0.8.32

EDIT: Has .32 been pushed yet? Latest on master 0.8.31

**Comment by tposney** [comment]
Arghh, I managed not to push 0.8.32 when I thought I had. Bother. I'll have to check the 0.8.33 changes I've got queued and then I'll push that version.

0.8.33 just pushed which should have the fix as well.

**Comment by tposney** [comment]
In GitLab by @nickelpro on Jul 12, 2021, 22:47

New but related error on 0.8.33. Same stage, same style of issue (trying to read from `null` item), new line:

```
workflow.js:642 Uncaught (in promise) TypeError: Cannot read property 'data' of null
    at workflow.js:642
    at Map.filter (collection.mjs:66)
    at DamageOnlyWorkflow._next (workflow.js:626)
    at DamageOnlyWorkflow._next (workflow.js:1727)
    at async DamageOnlyWorkflow.next (workflow.js:334)
```

which is
```javascript
  const abl = this.item.data.data.save.ability;
```

**Comment by tposney** [comment]
Can you send me the macro and actor that is doing the call? My tests are not causing the same problems.

**Comment by tposney** [comment]
In GitLab by @nickelpro on Jul 14, 2021, 01:41

Not getting it in a clean test world, must be some interaction on my end. Closing

**Comment by tposney** [comment]
In GitLab by @nickelpro on Jul 14, 2021, 02:01

Figured it out, the code makes it clear the conditions under which this can happen:

From workflows.ts

```javascript
case WORKFLOWSTATES.ROLLFINISHED:
    warn('Inside workflow.rollFINISHED');
    // Do special expiries
    for (let target of this.targets) {
        //@ts-ignore effects
        const expiredEffects = target.actor?.effects?.filter(ef => {
            const wasAttacked = this.item?.hasAttack;
            const wasDamaged = itemHasDamage(this.item) && this.applicationTargets?.has(target); //TODO this test will fail for damage only workflows - need to check the damage rolled instaed
            const specialDuration = getProperty(ef.data.flags, "dae.specialDuration");
            if (!specialDuration)
                return false;
            //TODO this is going to grab all the special damage types as well which is no good.
            if ((specialDuration.includes("isAttacked") && wasAttacked) ||
                (specialDuration.includes("isDamaged") && wasDamaged))
                return true;
            if ((specialDuration.includes("1Reaction")) && target.actor.uuid !== this.actor.uuid)
                return true;
            if (this.item?.hasSave && specialDuration.includes(`isSaveSuccess`) && this.saves.has(target))
                return true;
            if (this.item?.hasSave && specialDuration.includes(`isSaveFailure`) && !this.saves.has(target))
                return true;
            const abl = this.item.data.data.save.ability;
```

Inside a DamageOnlyWorkflow, when checking expired effects with a special duration, and none of the previous conditions have caused the callback to return this line:
`const abl = this.item.data.data.save.ability;`
Causes the error.

I don't know the answer here, because you would want effects that expire on saving throws to be able to be triggered by a damageOnly workflow, but the workflow has no way of knowing about the saving throws for an item it can't see.

**Comment by tposney** [comment]
In GitLab by @nickelpro on Jul 14, 2021, 02:08

To trigger this just put a saving-throw expired Active Effect on the target of a DamageOnlyWorkflow. My reproduction is literally just that but I'll attach anyway

[fvtt-Actor-attacker.json](https://my-gitlab-bucket.s3.amazonaws.com/640711674/c54722f957c84062ec36f0858e9edd154c04d9f79698b69d76b2f1bc74337948/fvtt-Actor-attacker.json)
[fvtt-Actor-target.json](https://my-gitlab-bucket.s3.amazonaws.com/640711674/0a695c7cd3723026d326c412fcf32194f78a1a600684eb74b067d0177bca946c/fvtt-Actor-target.json)

**Comment by tposney** [comment]
I've put in a fix for checking special duration expiry when there is no item in the workflow, but it feels a bit stop gap, but at least wont fall over, until the next one . Fix will go out in 0.8.34.
